contents/csharp/CSAsyncAwait.md

```diff
@@ -0,0 +1,322 @@
+<frontmatter>
+  title: Asynchronous Programming In C# Using Async/Await
+  header: pagetop.md
+  footer: footer.md
+  head: head.md
+  siteNav: mainNav.md
+  pageNav: 3
+</frontmatter>
+
+<div class="website-content">
+
+{{ booktitle | safe }}
+
+# Asynchronous Programming In C# Using Async/Await
+
+**Author: [Muhammad Irham Rasyidi Bin Zainal](https://github.com/xantho09)**<br>
+
+<box id="article-toc">
+
+* [What Is Asynchronous Programming](#what-is-asynchronous-programming)
+* [Example Of Asynchronous Programming](#example-of-asynchronous-programming)
+* [Basic Syntax‎](#basic-syntax)
+    * [Async](#async)
+    * [Await](#await)
+    * [Returning Values](#returning-values)
+* [Control Flow‎](#control-flow)
+* [Disadvantages Of Async/Await](#disadvantages-of-asyncawait)
+    * [Potential Pitfall: Deadlock](#potential-pitfall-deadlock)
+    * [Additional Overhead](#additional-overhead)
+* [Resources](#resources)
+* [Further Reading](#further-reading)
+
+</box>
+
+## What Is Asynchronous Programming?
+
+In **synchronous programming**, code is executed one step at a time, waiting for each step to be finished before proceeding to the next step. For example, if the code looks something like this:
+
+1. Perform Task 1.
+2. Perform Task 2.
+3. Perform Task 3.
+4. Use the result from Task 1 to perform Task 4.
+
+Then a synchronous execution of this code will start executing Task 1 and wait for it to finish before starting Task 2. Then it will wait for Task 2 to finish before starting Task 3, and so on.
+
+In general, this would be fine. However, if Task 1 happens to take a long time to finish, then a lot of time is spent in the execution simply waiting. We can make better use of our time by executing the code like this:
+
+1. Start running Task 1.
+2. Start running Task 2 _(While waiting for Task 1 to finish)_.
+3. Start running Task 3 _(While waiting for Task 1 to finish)_.
+4. Wait for Task 1 to finish.
+5. Use the result from Task 1 to start Task 4.
+6. Wait for Tasks 2, 3 and 4 to finish.
+
+This is the main idea behind asynchronous programming: while waiting for tasks to complete in the background, perform other tasks instead of just waiting there and doing nothing. This can result in better utilization of resources at hand.

```

Comment:
Well technically the current task is being executed. Is there a layman's way to explain execution contexts or just mention threading?

Reply:
Admittedly, the introduction was the hardest part to write. It was tricky to keep the explanation simple while making sure it gets the idea across.

I think mentioning threading is probably the simplest solution. Something along the lines of:
- You have 3 additional threads, A, B and C.
- Start Task 1 on Thread A.
- Start Task 2 on Thread B.
And so on.

Reply:
How does it look now?

I'm not too satisfied with the sentence that follows _"This is the main idea behind asynchronous programming"_; I feel like there's a better way to phrase it.